src/backend/distributed/transaction/transaction_management.c

```diff
@@ -317,12 +319,19 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 			MemoryContext previousContext =
 				MemoryContextSwitchTo(CitusXactCallbackContext);
 
-			if (CurrentCoordinatedTransactionState == COORD_TRANS_PREPARED)
+			if (CurrentCoordinatedTransactionState == COORD_TRANS_PREPARED &&
+				!IsMainDBCommand)
 			{
 				/* handles both already prepared and open transactions */
 				CoordinatedRemoteTransactionsCommit();
 			}
 
+			if (!IsCitusMainDB())
+			{
+				RunCitusMainDBQuery("SELECT commit_management_command_2PC()");
```

Comment:
In general, using some macros instead of hardcoded "SELECT udf" strings (to hardcode them once), where the UDF is schema qualified to avoid from the problems that might arise when the current search_path is not properly set on workers (due to a bug or something).